### ISSUE TITLE
Fragmentation?

### ISSUE THREAD

**Issue by tjconcept** [nodejs/readable-stream#115]
Hi streams people!

Streams is one of the ideas I love the most in Node, nonetheless I do not use them unless I'm forced to, and I keep them as isolated as possible. They've never actually caused me trouble and I understand the API and principles.
The reason is simple; when a new "tool" comes to my attention I read everything about it, evaluate the players and pick a horse. I just haven't been able to within reasonable time with streams.
There are so many packages mentioning one or more of streams1/2/3, node 0.8/0.10/0.12 streams, readable-stream 1.0.x/1.1.x that I have no clue how they play together.

It might already exist somewhere, but what I would really love is an overview of the different implementations, versions, their history, differences and how they play together.
There also seems to exist an enormous amount of packages which basically provide basic streams implementation like through, duplex, readable, etc. When should they be used? Are they redundant?

The concept of a stream is so elegant that it should be used everywhere, it just seems that too few are actually doing this, and they might have a similar reason to mine.

**Comment by calvinmetcalf** [comment]
tl;dr; just do `npm install --save readable-stream' and `var stream = require('readable-stream');`


currently the streams are all backwards compatible (mostly) with one of the big differences being streams 2 (aka node 0.10 streams aka readable-stream 1.0) is that it added the ability to deal with back pressure, in other words when you where reading from a stream and doing something async with it in streams 1 you had to manually call pause while you where processing it and hope the other stream respected that or else 

![backpressure](https://cloud.githubusercontent.com/assets/1128607/6413736/565c5d18-be60-11e4-8a75-799fdf88d9ff.gif)

there are some edge cases that can lead to incompatibility between streams1 and streams 2 [detailed here](https://iojs.org/api/stream.html#stream_compatibility_with_older_node_js_versions).

streams 3 aka 0.11/0.12 streams aka readable streams 1.1 involved some under the hood changes that I don't know off the top of my head and some new features like corking, but again you could use a stream 3 as a stream 2 with no issue.

The whole readable stream repo was originally created so you could avoid caring about these issues as detailed [here](http://r.va.gg/2014/06/why-i-dont-use-nodes-core-stream-module.html) so instead of having to worry about which version of node you're app might end up running on instead you just use readable-stream and have it be the same across versions of node.

**Comment by chrisdickinson** [comment]
> The whole readable stream repo was originally created so you could avoid caring about these issues as detailed here so instead of having to worry about which version of node you're app might end up running on instead you just use readable-stream and have it be the same across versions of node.

One thing to note: there's not unanimous agreement about whether one should use this package directly or use the streams built into your node runtime. A few (myself included) prefer that end users stick to the streams packages with the runtime, since pegging one's module to a given version of readable-stream can (at most) solve one half of the interoperability problem.

Opinions aside, though, you should be okay to `require('stream').Readable` OR `require('readable-stream').Readable` (or Duplex, Transform, Writable, etc.)

**Comment by tjconcept** [comment]
Thanks both, that was really clarifying. Why don't you add an explanation like that somewhere loud? :)

So, as long as I'm on node >= 0.10, compatibility shouldn't be a problem.
I've read about "corking", but that probably wouldn't be a good idea to use as long as I'm on 0.10, as it wouldn't be supported by core modules.
Is readable-stream the actual code used in iojs builds?

Conclusion; I should either use the core module or readable-stream (1.1) but avoid cork for node < 0.11 compatibility.

P.S. that GIF did not do anything good to the reading experience ;-)

**Comment by calvinmetcalf** [comment]
if you use readable stream 1.1 then you can safely use cork on streams you
create

**Comment by calvinmetcalf** [comment]
closing this, if you still have questions feel free to reopen
